File: algorithms/cpp/45.跳跃游戏-ii.cpp
#include <iostream>
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=45 lang=cpp
 *
 * [45] 跳跃游戏 II
 */

// @lc code=start
class Solution {
 public:
  int jump(vector<int>& nums) {
    int steps = 0, max_pos = 0, end = 0;
    for (int i = 0; i < nums.size() - 1; i++) {
      max_pos = max(max_pos, i + nums[i]);
      if (i == end) {
        end = max_pos;
        steps++;
      }
    }
    return steps;
  }
};
// @lc code=end

File: algorithms/golang/45.跳跃游戏-ii.go
/*
 * @lc app=leetcode.cn id=45 lang=golang
 *
 * [45] 跳跃游戏 II
 */

// @lc code=start
/**
 * 从后往前贪心，找出走最大步幅到达目的地的位置，该点即为边界点。
 */
func jump(nums []int) int {
	steps, position := 0, len(nums)-1
	for position > 0 {
		for i := 0; i < position; i++ {
			if nums[i]+i >= position {
				position = i
				steps++
			}
		}
	}
	return steps
}

// @lc code=end
